### PR TITLE
feat: remove `data.<query name>` from rest api responses

### LINKED ISSUE #1436: drop `data.<query name>` from API responses in the REST router
The REST request handler returns a response that looks as follows

**Current**
Wraps the actual values with levels.
```json5
{
  "data": {
     "<query name>": {foo: 1, bar: true}
  }
}
```

**Proposed**
Just simply return the data

```json5
{foo: 1, bar: true}
```

### PR DESCRIPTION
**Pull request by ologbonowiwi**
**Issue Reference(s):**  
Fixes #1436
/claim #1436

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

## Summary by CodeRabbit

- **New Features**
	- Enhanced response data handling in REST APIs with new parsing and manipulation capabilities.
- **Refactor**
	- Improved clarity in REST API response handling by renaming and updating functions and methods for better specificity and efficiency.
	- Streamlined GraphQL responses by introducing a method to flatten and transform data structures for RESTful integration.

### CODE REVIEW

File: src/http/request_handler.rs [hunk 1]
```diff
@@ -166,6 +168,32 @@ fn create_allowed_headers(headers: &HeaderMap, allowed: &BTreeSet<String>) -> He
...
+
+            if let serde_json::Value::Object(data) = body.data.take() {
+                let values: Vec<serde_json::Value> = data.values().cloned().collect();
+                if let Some(value) = values.first() {
+                    *resp.body_mut() = Body::from(value.to_string());
+                }
+            }
+        }
+    }
+}
```
Comment (ologbonowiwi) [inline src/http/request_handler.rs:195]: @coderabbitai create tests for this method

File: src/http/request_handler.rs [hunk 2]
```diff
@@ -187,6 +226,19 @@ async fn handle_rest_apis(
             app_ctx.as_ref(),
         );
         update_experimental_headers(&mut resp, app_ctx.as_ref(), req_ctx);
+
+        if let Some(body) = get_value_from_response(&mut resp).await {
```
Comment (tusharmath) [inline src/http/request_handler.rs:230]: We should perform this operation before converting the value to response. GraphQLResponse contains deserialized version of the ConstValue, on which we can quickly pattern match and extract out the things we don't need.

**Comment by tusharmath** [comment]
@ologbonowiwi Are you still on this?

**Comment by ologbonowiwi** [comment]
Yes @tusharmath. I'll push the last changes but the point I'm where now, the feature is pretty much done but I'm having issues to generate a `GraphQLResponse` which has non-empty `errors`.

Some of the approaches I've tried so far: GraphQL resolver returning error, HTTP resolver returning error, and returning status 500 on data source. Somewhere on the response conversion the data turns into null values, the error property keeps empty still.

File: src/async_graphql_hyper.rs [hunk 3]
```diff
@@ -128,6 +128,52 @@
...
+                let item = list
+                    .iter()
+                    .map(|res| Self::flatten_response(res.data.clone()))
+                    .collect::<Value>();
+                let data = serde_json::to_string(&item)?;
+
+                self.build_response(StatusCode::OK, Body::from(data))
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ologbonowiwi) [reply src/async_graphql_hyper.rs:175]: Yes. Can you generate tests to this?

File: src/async_graphql_hyper.rs [hunk 4]
```diff
@@ -128,6 +128,52 @@ impl GraphQLResponse {
...
+
+    fn flatten_response(data: Value) -> Value {
+        if let Value::Object(map) = data.clone() {
+            if map.len() == 1 {
+                if let Some(value) = map.values().next() {
+                    return value.clone();
+                }
+            }
+        }
+        data
```
Comment (ologbonowiwi) [inline src/async_graphql_hyper.rs:147]: That's unreachable. I'm keeping it here as fallback, but we should never reach this point.

File: src/async_graphql_hyper.rs [hunk 5]
```diff
@@ -128,6 +128,52 @@ impl GraphQLResponse {
...
+    }
+
+    pub fn to_response(self) -> Result<Response<hyper::Body>> {
+        self.build_response(StatusCode::OK, self.default_body()?)
+    }
+
+    fn flatten_response(data: Value) -> Value {
+        if let Value::Object(map) = data.clone() {
+            if map.len() == 1 {
+                if let Some(value) = map.values().next() {
```
Comment (tusharmath) [inline src/async_graphql_hyper.rs:142]: use `swap_remove` so that you can avoid that clone.

File: src/async_graphql_hyper.rs [hunk 6]
```diff
@@ -128,6 +128,52 @@ impl GraphQLResponse {
...
+            BatchResponse::Single(ref res) => {
+                let item = Self::flatten_response(res.data.clone());
+                let data = serde_json::to_string(&item)?;
+
+                self.build_response(StatusCode::OK, Body::from(data))
+            }
+            BatchResponse::Batch(ref list) => {
+                let item = list
+                    .iter()
+                    .map(|res| Self::flatten_response(res.data.clone()))
```
Comment (tusharmath) [inline src/async_graphql_hyper.rs:168]: avoid clone. This causes massive performance degradation.

File: src/async_graphql_hyper.rs [hunk 7]
```diff
@@ -128,6 +128,52 @@ impl GraphQLResponse {
...
+    /// Differs as `to_response` by flattening the response's data
+    /// `{"data": {"user": {"name": "John"}}}` becomes `{"name": "John"}`.
+    pub fn to_rest_response(self) -> Result<Response<hyper::Body>> {
+        if !self.0.is_ok() {
+            return self.build_response(StatusCode::INTERNAL_SERVER_ERROR, self.default_body()?);
+        }
+
+        match self.0 {
+            BatchResponse::Single(ref res) => {
+                let item = Self::flatten_response(res.data.clone());
```
Comment (tusharmath) [inline src/async_graphql_hyper.rs:160]: avoid clone here. This causes massive performance degradation

**Review by tusharmath — APPROVED** [review]

**Pull request merged**